### PR TITLE
Implementing thumbnail rendering support for cases where Figgy GraphQL responses do not contain the bib. IDs

### LINKED ISSUE #2411: Thumbnails are broken for figgy images
https://catalog-alma-qa.princeton.edu/catalog/9946093213506421
https://catalog-alma-qa.princeton.edu/catalog/9947652213506421

This is because we changed figgy so that if you hit for the graphql endpoint it looks both versions of the id. But it returns whichever version of the ID it found in the response, so if the id isn't the same in figgy then the match-up for placing the thumbnail doesn’t work.

Implementation: look for both versions of the ID on the javascript side when placing the thumbnail.

### PR DESCRIPTION
**Pull request by jrgriffiniii**
Resolves #2411 and addresses the following:

- Introduces `capybara-screenshot` as a Gem dependency (screenshots for failing test suites were not being generated in my local development environment)
- Introduces a `selenium_headless` configuration for Capybara (this was not headless in my local environment)
- Introduces system specs for in-browser testing (please see https://relishapp.com/rspec/rspec-rails/docs/system-specs/system-spec)

### CODE REVIEW

**Review by bess — APPROVED** [review]

File: Gemfile [hunk 1]
```diff
@@ -74,7 +74,7 @@ gem 'stringex', git: 'https://github.com/pulibrary/stringex.git', tag: 'vpton.2.
 gem 'mail_form'
 gem 'string_rtl'
 
-gem 'requests', git: 'https://github.com/pulibrary/requests.git', branch: 'alma'
+gem 'requests', git: 'https://github.com/pulibrary/requests.git', branch: 'main'
```
Comment (hackartisan) [inline Gemfile:77]: I don't think we want to point back to main. we will lose all the requests updates

**Review by hackartisan — CHANGES_REQUESTED** [review]

File: config/application.rb [hunk 2]
```diff
@@ -45,7 +45,5 @@ class Application < Rails::Application
     config.robots = OpenStruct.new(config_for(:robots))
 
     config.use_alma = ActiveModel::Type::Boolean.new.cast(ENV['USE_ALMA'])
-
-    config.alma = config_for(:alma).with_indifferent_access
```
Comment (hackartisan) [inline config/application.rb:49]: Is this related to this PR?

File: config/requests.yml [hunk 3]
```diff
@@ -27,7 +27,6 @@ staging:
 alma_qa:
   <<: *defaults
   bibdata_base: https://bibdata-alma-staging.princeton.edu
-  scsb_base: https://uat-recap.htcinc.com:9093
```
Comment (hackartisan) [inline config/requests.yml:30]: why is this deleted?

File: config/application.rb (hunk 2 above)
Comment (hackartisan) [reply config/application.rb:49]: also config/alma.yml is removed... I wonder if this needs to be rebased against the alma branch.

**Review by hackartisan — CHANGES_REQUESTED** [review]
I'm not really sure I follow the PR since I think there are unrelated changes. I think rebasing to the alma branch will fix it and then I'd like to look at it again.

File: Gemfile (hunk 1 above)
Comment (jrgriffiniii) [reply Gemfile:77]: That's strange, I fixed this last week, I am sorry, I will see how this regressed into my branch.

Comment (hackartisan) [reply Gemfile]: the rebase fixed it.

**Comment by jrgriffiniii** [comment]
This needs to be implemented to support regenerating the Bib. ID from the Voyager ID (the latter is returned from the GraphQL response, and the client can simply prefix and append a constant suffix to generate this consistently).

File: app/services/bibdata.rb [hunk 4]
```diff
@@ -53,9 +53,18 @@ def holding_locations
 
     private
 
+      def parse_holding_locations(response:)
```
Comment (jrgriffiniii) [inline app/services/bibdata.rb:56]: These were introduced due to errors which I encountered within my development environment (and on CircleCI).

File: spec/support/capybara_selenium.rb [hunk 5]
```diff
@@ -1,21 +1,26 @@
...
   Capybara::Selenium::Driver.new(app, browser: :chrome)
 end
+
 Capybara.register_driver(:selenium) do |app|
   capabilities = Selenium::WebDriver::Remote::Capabilities.chrome(
     chromeOptions: { args: %w[headless disable-gpu disable-setuid-sandbox window-size=7680,4320] }
   )
+
   browser_options = ::Selenium::WebDriver::Chrome::Options.new
   browser_options.args << "--headless"
```
Comment (jrgriffiniii) [inline spec/support/capybara_selenium.rb:17]: This might not be necessary, I couldn't manage to get this to run headlessly for my local dev. environment (nor in CircleCI).

File: app/services/bibdata.rb [hunk 6]
```diff
@@ -53,9 +53,18 @@ def holding_locations
...
+        JSON.parse(response.body)
+      rescue JSON::ParserError => parse_error
+        Rails.logger.warn("Unable to parse response from the locations endpoint: #{parse_error}")
+        []
+      end
+
       def sorted_locations(response)
         locations_hash = {}.with_indifferent_access
-        JSON.parse(response.body).each do |location|
+        holding_locations_response = parse_holding_locations(response: response)
```
Comment (jrgriffiniii) [inline app/services/bibdata.rb:65]: This should be removed, as this was resolved upstream in the `requests` app.

File: spec/system/catalog_show_spec.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+      expect(page).to have_selector ".document-thumbnail.has-viewer-link"
+      node = page.find(".document-thumbnail.has-viewer-link")
+      expect(node["data-bib-id"]).not_to be_empty
+      expect(node["data-bib-id"]).to eq document_id
+    end
+
+    context 'when the Figgy GraphQL responds with a different bib. ID' do
+      let(:document_id) do
+        '9947652213506421'
```
Comment (hackartisan) [inline spec/system/catalog_show_spec.rb:53]: I expected this test to use id 4609321

File: spec/features/request_options_spec.rb [hunk 8]
```diff
@@ -3,7 +3,32 @@
 require 'rails_helper'
 
 describe 'Request Options' do
-  before { stub_holding_locations }
+  let(:availability_response) do
```
Comment (hackartisan) [inline spec/features/request_options_spec.rb:6]: Did you try the tests again without the changes to this file?

File: spec/support/capybara_selenium.rb [hunk 9]
```diff
@@ -1,32 +1,53 @@
...
+
+  Capybara::Selenium::Driver.new(app,
+                                 browser: :chrome,
+                                 desired_capabilities: capabilities,
+                                 http_client: http_client,
+                                 options: browser_options)
+end
+
+# This was needed for my local workstation, perhaps :selenium is overridden elsewhere?
+Capybara.register_driver(:selenium_headless) do |app|
```
Comment (hackartisan) [inline spec/support/capybara_selenium.rb:31]: 👍🏻

**Review by hackartisan — CHANGES_REQUESTED** [review]

**Review by hackartisan — APPROVED** [review]
